crates/runtime/src/traphandlers.rs

```diff
@@ -397,3 +429,120 @@ mod tls {
         })
     }
 }
+
+/// A module for registering a custom alternate signal stack (sigaltstack).
+///
+/// Rust's libstd installs an alternate stack with size `SIGSTKSZ`, which is not always
+/// large enough for our signal handling code, and which as of this writing does not
+/// have a guard page. Override it by creating and registering our own alternate stack
+/// that is large enough and has a guard page.
+#[cfg(unix)]
+mod unix_sigaltstack {
+    use std::cell::RefCell;
+    use std::convert::TryInto;
+    use std::ptr::null_mut;
+
+    thread_local! {
+        /// Thread-local state is lazy-initialized on the first time it's used, and dropped
+        /// when the thread exits.
+        pub(crate) static TLS: RefCell<Tls> = RefCell::new(unsafe { Tls::new() });
+    }
+
+    /// The size of the sigaltstack (not including the guard, which will be added).
+    /// Make this large enough to run our signal handlers.
+    static STACK_SIZE: usize = libc::SIGSTKSZ * 4;
+
+    /// A utility to register a new sigaltstack.
+    pub(crate) struct Tls {
+        guard_size: usize,
+        sigaltstack_alloc_size: usize,
+        new_stack: libc::stack_t,
+    }
+
+    impl Tls {
+        unsafe fn new() -> Self {
+            let page_size: usize = libc::sysconf(libc::_SC_PAGESIZE).try_into().unwrap();
+            let guard_size = page_size;
+
+            let mut me = Self {
+                guard_size,
+                sigaltstack_alloc_size: guard_size + STACK_SIZE,
+                new_stack: libc::stack_t {
+                    ss_sp: null_mut(),
+                    ss_flags: 0,
+                    ss_size: 0,
+                },
+            };
+
+            // Allocate memory.
+            let ptr = libc::mmap(
+                null_mut(),
+                me.sigaltstack_alloc_size,
+                libc::PROT_NONE,
+                libc::MAP_PRIVATE | libc::MAP_ANON,
+                -1,
+                0,
+            );
+            if ptr == libc::MAP_FAILED {
+                super::raise_oom_trap()
```

Comment:
Instead of raising a trap could this perhaps bubble up some sort of error to the caller? I don't think the raising mechanism is needed here if this is reorganized a bit, for example the caller could do:

```rust
#[cfg(unix)]
unix::setup_sigaltstack()?;
```

and internally this module would check the TLS variable, and if it's not initialized would perform inline initialization which would bubble this error up through returns rather than a longjmp

Reply:
I might recommend something like:

```rust
thread_local!(static TLS: RefCell<Tls> = RefCell::new(Tls::None);

enum Tls {
    None,
    Allocated(Sigaltstack), // `Drop for Sigaltstack` frees stuff
    CheckedAlready), // sigaltstack was previously big enough
}
```